src/ripple/protocol/impl/AccountID.cpp

```diff
@@ -25,9 +25,135 @@
 
 namespace ripple {
 
+#ifndef RIPPLED_ACCOUNTID_CACHE_SIZE
+#define RIPPLED_ACCOUNTID_CACHE_SIZE 1000000
+#endif
+
+namespace detail {
+
+/** Caches the base58 representations of AccountIDs */
+class AccountIdCache
+{
+private:
+    // The total number of items the cache buffer can contain
+    std::size_t const count_;
+
+    // The maximum size (in bytes) of an item, along with a null terminator
+    // and aligned at 8 bytes
+    std::size_t const size_ = 40;
+
+    // The offset of the next available block of `size_` contiguous bytes
+    std::size_t index_ = 0;
+
+    // The memory block
+    std::unique_ptr<char[]> chars_;
+
+    // The map
+    hardened_hash_map<AccountID, char*> map_;
+
+    std::mutex mutable mutex_;
+
+    // The total number of requests to format an AccountID
+    std::size_t requests_ = 0;
+
+    // The number of times were able to satisfy the request from cache
+    std::size_t hits_ = 0;
+
+public:
+    AccountIdCache(std::size_t count)
+        : count_(count)
+        , chars_(std::make_unique<char[]>(count_ * size_))
+        , map_(count_ / 971)
+    {
+        std::memset(chars_.get(), 0, count_ * size_);
+    }
+
+    std::string
+    toBase58(AccountID const& id)
+    {
+        assert(chars_.get() != nullptr);
+
+        std::lock_guard lock(mutex_);
```

Comment:
My biggest concern with this cache is lock contention. We know that base58 encoding is expensive. On a cache miss this change does that encoding under a lock. 

What do you think about using a boost `UpgradeLock` here? We could check out a read lock, and on a cache miss, do the base58 encoding _before_ we upgrade the lock and write to the cache. That might help with lock contention. It might be a good idea to write some simple benchmarks to experiment with such changes.

Reply:
The previous cache did the same thing, but I agree that avoiding doing expensive work (like encoding) under the lock makes sense. My latest changes should address this in two ways: (a) not holding a lock while encoding; and (b) carefully allowing concurrent access to the cache.

I'm marking this as resolved, but feel free to undo this if you think it's still an issue.